### DISCUSSION TITLE
ESP32 system failure during deepsleep outcome?

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#10498] [ESP32] [answered]
I want to use deepsleep on a project and was hoping someone has some 
real-world experience with the ESP32.

I currently used lightsleep on a ESPNow link, wake up every minute and pat 
both a system WDT and a SW WDT.  I now want to put the system into deepsleep
for an hour and as the units will re-boot, via a RTC alarm, any watchdogs
now seem irrelevant.

Correct me, if I have this wrong.

What will happen to an ESP32 in deepsleep if something crashes somewhere, will
it re-boot or just stop working?

**Comment by karfas** [comment, marked as answer]
> What will happen to an ESP32 in deepsleep if something crashes somewhere, will it re-boot or just stop working?

This question is too broad. When your power supply "crashes", the ESP32 will stop working too (of course).

There are only minor differences between a hard reset and coming out of deep sleep.
In my opinion, an ESP32 not waking up from a timed deep sleep is the least thing you need to worry about when designing a high-available system.
Even that can be fixed by an external watchdog doing a reset after two hours or so - at the cost of some microamperes for the watchdog.

However, you need to ensure (by your software watchdog, perhaps) that the ESP32 really enters deep sleep after your program did it's thing.
You don't want a perfectly working ESP32 sitting there and waiting for REPL input...

**Reply by davefes** [reply]
"crashes somewhere" I should have said in the operating system (vitural machine or FreeRTOS).  The power is from a single Li-ion battery and LDO.  The hardware I trust!

> However, you need to ensure (by your software watchdog, perhaps) that the ESP32 really enters deep sleep after your program did it's thing.  You don't want a perfectly working ESP32 sitting there and waiting for REPL input...

I don't understand why I need to check if it has really gone to sleep and what REPL has got to do with it,  Could you please elaborate?

**Reply by karfas** [reply]
This discussion is purely academic, I think. By creating the application, you already decided to "trust" the processor, the RTOS, the software as well as your power supply. Deep sleep is a core feature of the ESP32; there is no RTOS involved during deep sleep.

In absence of (some) watchdog, e.g. any unhandled exception will interrupt your program and give you a nice REPL prompt (and therefore no way to enter deep sleep or reboot).

**Reply by davefes** [reply]
Over the 2 years I have been using the ESP32 unexplained crashes have virtually disappeared, probably because I have learnt a few things about writing proper code!  However, when making changes, I am always expecting that something new is going to catch me unawares.

I will test using, say 90 minute watchdog periods for the system and software watchdogs during deepsleep.

From the docs:

> The ESP32 port has four hardware timers. Use the [machine.Timer](https://docs.micropython.org/en/latest/library/machine.Timer.html#machine-timer) class with a timer ID from 0 to 3 (inclusive)

I got the impression that these timers used a different clock than the RTC function, but reading several other articles, ie:
https://lastminuteengineers.com/esp32-deep-sleep-wakeup-sources/

> The ESP32 RTC controller has a built-in timer that you can use to wake up the ESP32 after a predefined amount of time.

it now seems that these four hardware timers will the have poor timing accuracy of the RTC clock.

Your comments were helpful.  Thank you.

**Reply by robert-hh** [reply]
> poor timing accuracy of the RTC clock

Glancing through the schematics of a few common ESP32 boards, none of them has a 32k crystal. Which explains why the RTC accuracy is poor. The espressif modules have the pins for the 32k crystal exposed, but board vendors use this pin for other purposes.

**Reply by davefes** [reply]
I did consider a 32kHz xtal but came to the conclusion that I would need to develop my skills at rebuilding the firmware.

Should the docs make it clearer that these four hardware timers are RTC timers?

**Reply by robert-hh** [reply]
According to the code, the timer module uses the espressif HW timers, and these are clocked by the APB clock, which is 80 MHz by default and derived from a crystal, usually 40MHz.

**Reply by davefes** [reply]
If you go to deepsleep for 60 minutes then that uses the RTC timer, correct?

If I go to deepsleep and use my software wdt that uses machine.timer then it will wake-up at 60 minutes +/- crystal frequency tolerance, correct?

If so, then I don't need to use a DS3231 with a one hour alarm to wake-up the system.

@robert-hh thanks for your added comments.

**Reply by karfas** [reply]
I don't think that the four hardware timers called by machine.Timer() have anything to do with the RTC or the RTC clock.
From ESP-IDF documentation:
> RTC controller has a built in timer which can be used to wake up the chip after a predefined amount of time. Time is specified at microsecond precision, but the actual resolution depends on the clock source selected for RTC SLOW_CLK.

Some details to the RTC timer you find in https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#ulp . 
The **default** clock source **during deep sleep** for the RTC clock is the internal 150kHz timer, but it looks like other clock sources can be configured (at cost of leaving some parts of the chip powered on).

**Reply by davefes** [reply]
@karfas  I naively thought that there would be a power consumption cost running these four timers.  When most articles state "everything is shut-down except for the RTC timer" that suggests to me that using these four timers are going to incur extra power consumption.

Heh, if I can keep the total deepsleep current under 1mA I don't need dig any deeper!

Thanks, again.

**Reply by karfas** [reply]
I have no idea how `machine.Timer()` came into this discussion.

@davefes: it would be "*naive*" to assume that additional *things* will be working without power consumption.

> Heh, if I can keep the total deepsleep current under 1mA I don't need dig any deeper!

I'm sure this is possible, but be aware that most ESP32 development boards were not designed with "low power" in mind. 
Components like the UART/USB converter, LEDs, ... might consume precious  microamperes, even when the ESP32 module itself runs with less than 10 uA during deep sleep.

**Reply by davefes** [reply]
I brought it in by mentioning that my SW WDT uses machine.timer().  If machine.timer stops running in deepsleep()  then this won't work.
Likewise, if machine.WDT stops running in deepsleep() then this won't work either.

I'll be able to do some tests in 2 days.  All the boards I use in low-power apps have the on-board LED and 3V3 regulator removed and a jumper link put in for the CP2102 UART bridge supply.

**Reply by karfas** [reply]
> If machine.timer stops running in deepsleep() then this won't work. Likewise, if machine.WDT stops running in deepsleep() then this won't work either.

You still don't realize what deep sleep is. *Only* some functions of the RTC module and the ULP processor are running; everything else gets no power. What should a watchdog do if a program that **is not supposed** to run (during deep sleep) really isn't running ?

Of course any kind of watchdog will be useful **while the application runs** so you can ensure the processor starts again from a clean state in case something unexpected happens.

**Reply by davefes** [reply]
> You still don't realize what deep sleep is.

Yes, I will admit that I am struggling a bit here.

I know that the program does not run during light or deepsleep.  I know that you can use the RTC timer to wake from deepsleep after a rough amount of time.  I see that ability to wake-up again important and also thought that it might be subject to failure.  I am probably confusing things by calling this ability to wake-up and start over again a sort of "watchdog".

I have some DS3231s coming, which I will set up to give one hour alarms to wake the ESP32s up from deepsleep.  I need timing accuracy as I need to sync three ESP32s, ie within a few seconds.

I have read quite a few articles on watchdogs and it does get quite complex the issues that people try to cater for, ie race conditions in code and cosmic rays!

**Comment by karfas** [comment]
>  I see that ability to wake-up again important and also thought that it might be subject to failure.

OK, but there is very little you can do about a not working or not booting device without external supervision. Most likely, any error during boot after deep sleep will also prevent a clean boot by machine.reset() and the like.
You will have to test and trust the ESP32 wakeup functions as well as you trust the reset button, the watchdog, and machine.reset().
The DS3231 will not help much in this regard.

If it is really that important (e.g. live-saving) that the device comes out from deep sleep, you could add one or more ESP32 and work out some logic so at least one of them performs it's function.

**Reply by davefes** [reply]
The main purpose of the DS3231 is to sync wake-up times for multiple devices so that I can establish two-way communications then go to sleep for an hour.  

Fortunately, this is not a life-saving situation. When I don't get an hourly update then it would mean a walk up some small hills to reset devices. 

I read through several of your PRs and appreciate the level of your comments.
